// nodejs-metrics/lib/external-service-metrics-tracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalServiceMetricsTracker = void 0;
const constants_1 = require("./constants");
const metrics_tracker_1 = require("./metrics-tracker");
class ExternalServiceMetricsTracker {
    constructor({ promClient }) {
        if (!promClient) {
            throw new Error("promClient argument is mandatory.");
        }
        const externalServiceLabels = Object.values(constants_1.ExternalServiceMetricConstants.Labels);
        const metrics = {
            [constants_1.ExternalServiceMetricConstants.Name]: new promClient.Histogram({
                name: constants_1.ExternalServiceMetricConstants.Name,
                help: `duration histogram of external service calls labeled with: ${externalServiceLabels.join(", ")}`,
                labelNames: externalServiceLabels,
                buckets: constants_1.ExternalServiceMetricConstants.HistogramValues.Bucket,
            }),
        };
        this.metricsTracker = new metrics_tracker_1.MetricsTracker({ metrics });
    }
    async trackHistogramDuration({ targetLabel, action, handleResult, }) {
        return await this.metricsTracker.trackHistogramDuration({
            metricName: constants_1.ExternalServiceMetricConstants.Name,
            labels: {
                [constants_1.ExternalServiceMetricConstants.Labels.Target]: targetLabel,
            },
            action,
            handleResult,
        });
    }
}
exports.ExternalServiceMetricsTracker = ExternalServiceMetricsTracker;
